### PR TITLE
SIP-100 Resolver & Cache Improvements

### PR DESCRIPTION
**Pull request by justinjmoses**
**Goals**
1. To reduce the load on deploys - processing many `setResolverAndSyncCache` calls is simply not scalable. 
1. To minimize the use of superuser owner functionality
1. To reduce deployment sizes of contracts by reducing constructor logic (in particular for the OVM deployment of contracts)

**Solidity Improvements**
* `AddressResolver.importAddresses()` emits an event `AddressImported` for each address added
* `AddressResolver.rebuildCaches` is exposed (and non-protected), for ease-of use
* `MixinResolver` no longer requires `Owned`. `resolver` is set once only on construction (with the expectation that `ReadProxyAddressResolver` will be used. 
* `MixinResolver` has a predicate view `isResolverCached()`  
* `MixinResolver` has a `rebuildCache()` function (publicly available) for the contract to rebuild its cache. This function also emits a `CacheUpdated` event for each item added to it's private cache
* `MixinResolver.requireAndGetAddress()` uses `abi.encodePacked` for string concatenation and a simpler API.
* `MixinResolver` no longer performs tasks in the constructor. Contracts using the mixin must implement the abstract `resolverAddressesRequired` function.
* `MixinSystemSettings` now wraps `MixinResolver` and returns the default `resolverAddressesRequired()` containing `FlexibleStorage`

**Stats**
* `AddressResolver.importAddresses(all 140 local targets)` takes approximately **3.5m gas**
* `AddressResolver.rebuildCaches(47 targets - most of the synths)` takes approximately **6.5m gas**

**Known Limitations**
1. As before, the deploy script will skip any contract with a dependency that has a `:` inside it. This is for the OVM bridge contracts, and `connect-bridge` must be run

### CODE REVIEW

**Review by ajsantander — COMMENTED** [review]
Looking good! Not only does this allow to sync caches in a single call, but it also feels safer and more efficient overall.

Left a few comments, nothing major.

File: contracts/AddressResolver.sol [hunk 1]
```diff
@@ -13,18 +15,28 @@ contract AddressResolver is Owned, IAddressResolver {
...
+            emit AddressImported(names[i], destinations[i], success);
         }
     }
 
-    /* ========== VIEWS ========== */
+    /* ========== PUBLIC FUNCTIONS ========== */
+    function rebuildCaches(MixinResolver[] calldata destinations) external {
+        // then, attempt to update all caches as required
+        for (uint i = 0; i < destinations.length; i++) {
+            destinations[i].rebuildCache();
```
[the remark this answers is not in the document]
Comment (justinjmoses) [reply contracts/AddressResolver.sol:35]: Yes - the idea is that the deploy script only invokes this with those contracts which expose the `rebuildCache()` function. I suppose we could accept any using a low-level call, but this seemed a reasonable approach. Thoughts?

File: contracts/MixinResolver.sol [hunk 2]
```diff
@@ -91,4 +74,19 @@ contract MixinResolver is Owned {
         // check the address exists already in the resolver
         addressCache[name] = resolver.getAddress(name);
     }
+
+    function requireAndGetAddress(bytes32 name) internal view returns (address) {
+        require(isCached, "Contract cache needs rebuilding");
+        address _foundAddress = addressCache[name];
+        require(_foundAddress != address(0), string(abi.encodePacked("Missing ", name, " address")));
```
[the remark this answers is not in the document]
Comment (justinjmoses) [reply contracts/MixinResolver.sol:81]: yup

File: contracts/AddressResolver.sol [hunk 3]
```diff
@@ -13,18 +15,28 @@ contract AddressResolver is Owned, IAddressResolver {
...
+    /* ========== RESTRICTED FUNCTIONS ========== */
 
     function importAddresses(bytes32[] calldata names, address[] calldata destinations) external onlyOwner {
         require(names.length == destinations.length, "Input lengths must match");
 
         for (uint i = 0; i < names.length; i++) {
             repository[names[i]] = destinations[i];
+            // solhint-disable avoid-low-level-calls
+            (bool success, ) = address(destinations[i]).call(abi.encodePacked(MixinResolver(0).invalidateCache.selector));
+            emit AddressImported(names[i], destinations[i], success);
```
[the remark this answers is not in the document]
Comment (justinjmoses) [reply contracts/AddressResolver.sol:27]: I figure it's useful information to have when searching for updates to a specific contract (especially for The Graph indexing) and since it's an upgrade action, the excess gas isn't a big deal

File: contracts/MixinResolver.sol [hunk 4]
```diff
@@ -34,44 +34,27 @@ contract MixinResolver is Owned {
         // Do not sync the cache as addresses may not be in the resolver yet
     }
 
-    /* ========== SETTERS ========== */
-    function setResolverAndSyncCache(AddressResolver _resolver) external onlyOwner {
```
[the remark this answers is not in the document]
Comment (justinjmoses) [reply contracts/MixinResolver.sol:38]: This is the advantage of using the `ReadProxyAddressResolver` - we no longer need to update it and we can remove the `Owned` requirement of every contract that uses it. `FlexibleStorage` for instance has no way to update it.

File: contracts/AddressResolver.sol [hunk 5]
```diff
@@ -13,18 +15,28 @@ contract AddressResolver is Owned, IAddressResolver {
...
         }
     }
 
-    /* ========== VIEWS ========== */
+    /* ========== PUBLIC FUNCTIONS ========== */
+    function rebuildCaches(MixinResolver[] calldata destinations) external {
```
[the remark this answers is not in the document]
Comment (justinjmoses) [reply contracts/AddressResolver.sol:32]: Any extra functionality here means it ends up in every contract that uses the mixin, so I'm trying to limit it. `isCached` is public, so the only extra functionality is to explicitly check the caching, which is an issue if `importAddresses` is called with only some of the contracts. For instance, if you call it with a single contract - like `Exchanger` - than any contract which needs `Exchanger` - say `Synthetix` - will not be invalidated. Hmmm, 🤔 let me marinate on this a bit more.

File: contracts/AddressResolver.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (justinjmoses) [reply contracts/AddressResolver.sol:35]: Actually scratch that - the danger with a low level call is that we are ignoring any reverts, which potentially misses a pretty important issue when rebuilding caches that we expect all the targets to already be in the resolver.  So I think this implementation is a better fit.

**Comment by justinjmoses** [comment]
Results from mainnet fork testing:

Baseline gas (when re-deploying contracts with no changes)
image

With the above changes
image

File: contracts/AddressResolver.sol [hunk 6]
```diff
@@ -13,17 +15,41 @@ contract AddressResolver is Owned, IAddressResolver {
...
+        // now rebuild the caches of all destinations if required
+        for (uint i = 0; i < destinations.length; i++) {
+            // solhint-disable avoid-low-level-calls
+            (bool success, ) = address(destinations[i]).call(abi.encodePacked(MixinResolver(0).rebuildCache.selector));
+            emit AddressImported(names[i], destinations[i], success);
+        }
+    }
+
+    /* ========= PUBLIC FUNCTIONS ========== */
+    function rebuildCaches(address[] calldata destinations) external {
```
Comment (ajsantander) [inline contracts/AddressResolver.sol:36]: This could be a bit sensitive. If an active contract is non included in destinations, and it has a module dependency of a contract that was upgraded, its cache will not be rebuilt.

Comment (justinjmoses) [reply contracts/AddressResolver.sol:36]: Yeah it's not perfect - it's similar to what we have right now with `setResolverAndSyncCache` - but I put a slew of checks in the deployment file for this.

File: contracts/AddressResolver.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (justinjmoses) [reply contracts/AddressResolver.sol:33]: Actually I can add this - it's just to the `AddressResolver` after all...

**Review by ajsantander — APPROVED** [review]
Looking great.

As noted in a comment, my only concern is that this new method can result in a partially cached system. I'd like to know what we're planning to do to avoid that.

File: contracts/MixinResolver.sol [hunk 7]
```diff
@@ -5,64 +5,63 @@ import "./Owned.sol";
...
+        resolver = AddressResolver(_resolver);
+    }
 
-    uint public constant MAX_ADDRESSES_FROM_RESOLVER = 24;
+    /* ========== INTERNAL FUNCTIONS ========== */
 
-    constructor(address _resolver, bytes32[MAX_ADDRESSES_FROM_RESOLVER] memory _addressesToCache) internal {
-        // This contract is abstract, and thus cannot be instantiated directly
-        require(owner != address(0), "Owner must be set");
+    function combineArrays(bytes32[] memory first, bytes32[] memory second)
```
Comment (i-stam) [inline contracts/MixinResolver.sol:23]: This one is used only in two contracts and seems a bit unrelated to MixinResolver

**Review by i-stam — APPROVED** [review]
lgtm, (just left a comment)

File: contracts/MixinResolver.sol (hunk 7 above)
Comment (justinjmoses) [reply contracts/MixinResolver.sol:23]: It's actually used in all contracts that extend from `MixinResolver` that want to add to the list of `resolverAddressesRequired`

**Comment by justinjmoses** [comment]
Replaced by #951
